src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -559,10 +567,22 @@ QString QgsPostgresUtils::whereClause( QgsFeatureId featureId, const QgsFields &
       whereClause = QStringLiteral( "%1=%2" ).arg( QgsPostgresConn::quotedIdentifier( fields.at( pkAttrs[0] ).name() ) ).arg( FID2PKINT( featureId ) );
       break;
 
+    case PktInt64:
     case PktUint64:
+    {
       Q_ASSERT( pkAttrs.size() == 1 );
-      whereClause = QStringLiteral( "%1=%2" ).arg( QgsPostgresConn::quotedIdentifier( fields.at( pkAttrs[0] ).name() ) ).arg( featureId );
-      break;
+      QVariantList pkVals = sharedData->lookupKey( featureId );
+      if ( !pkVals.isEmpty() )
+      {
+        QgsField fld = fields.at( pkAttrs[0] );
+        whereClause = conn->fieldExpression( fld );
+        if ( !pkVals[0].isNull() )
+          whereClause += "=" + pkVals[0].toString();
```

Comment:
This should be more efficient. Not completely sure if it works. If not, go for `QStringLiteral`

```suggestion
          whereClause +=  '=' + pkVals[0].toString();
```

Reply:
Done, see 9607570

Reply:
It does work.